ESCPOS_NET/Printers/BasePrinter.cs

```diff
@@ -1,51 +1,71 @@
+		private readonly int _maxBytesPerWrite = 15000; // max byte chunks to write at once.
+		
+		public PrinterStatusEventArgs Status { get; private set; } = null;
         public event EventHandler StatusChanged;
-        protected BinaryWriter _writer;
-        protected BinaryReader _reader;
-        protected System.Timers.Timer _flushTimer;
-        protected Thread _readThread;
-        protected ConcurrentQueue<byte> _readBuffer = new ConcurrentQueue<byte>();
-        protected int _bytesWrittenSinceLastFlush = 0;
-        private readonly int _maxBytesPerWrite = 15000; // max byte chunks to write at once.
-
-        public BasePrinter()
+        protected BinaryWriter Writer { get; set; }
+        protected BinaryReader Reader { get; set; }
+        protected System.Timers.Timer FlushTimer { get; set; }
+        protected ConcurrentQueue<byte> ReadBuffer { get; set; } = new ConcurrentQueue<byte>();
+        protected int BytesWrittenSinceLastFlush { get; set; } = 0;
+
+        protected BasePrinter()
         {
-            _flushTimer = new System.Timers.Timer(50);
-            _flushTimer.Elapsed += Flush;
-            _flushTimer.AutoReset = false;
+            FlushTimer = new System.Timers.Timer(50);
+            FlushTimer.Elapsed += Flush;
+            FlushTimer.AutoReset = false;
         }
+
         public virtual void Read()
         {
-            while (true)
+            while (_isMonitoring)
             {
                 try
                 {
-                    if (_monitoring)
-                    {
-                        // Sometimes the serial port lib will throw an exception and read past the end of the queue if a
-                        // status changes while data is being written.  We just ignore these bytes.
-                        var b = _reader.ReadByte();
-                        _readBuffer.Enqueue(b);
-                        DataAvailable();
+                    if (_cancellationTokenSource != null && _cancellationTokenSource.IsCancellationRequested)
+                    {
+                        _cancellationTokenSource.Token.ThrowIfCancellationRequested();
                     }
+
+                    // Sometimes the serial port lib will throw an exception and read past the end of the queue if a
+                    // status changes while data is being written.  We just ignore these bytes.
+                    var b = Reader.ReadByte();
+                    ReadBuffer.Enqueue(b);
+                    DataAvailable(false);
+                }
+                catch (OperationCanceledException ex)
+                {
+                    _cancellationTokenSource.Dispose();
+                    _cancellationTokenSource = null;
+                    _isMonitoring = false;
+                    Debug.WriteLine($"Read Cancelled Exception: {ex.Message}");
                 }
-                catch
+                catch (IOException ex)
                 {
-                    Thread.Sleep(100);
```

Comment:
this sleep was here to prevent this tight loop from using excessive CPU when no data reads were present, is the new version blocking?

Reply:
i think here (`IOException` and next `Exception ex`) we should set `_isMonitoring = false;` to end the task => also we need to reconnect to the printer(set to disconnected), as IOException throws when the TCP-connection is closed by the printer. 

however I'm not sure if this applies to all printer-types, this might be different for Serial/USB-printers?

Reply:
I believe serial printers also disconnect, I will verify this tomorrow.

Reply:
Serial printer don't disconnect, so it is isolated to IP printers only.

Reply:
I've added some re-connection code, maybe not the best  way to achieve it but it's working on the IP printer I tested on. We may want to make this optional as the disconnection happens for a reason.